### PR TITLE
Initial support for auth type selector in puppet test app

### PR DESCRIPTION
**Pull request by micarls**
* Add a new UI element which is a picker for authtype (currently B2C and AAD)
* Default to B2C
* Popup window telling user to restart app if auth type is changed
* Use different set of app secrets depending on auth type selected (guids for AAD empty for now, will fill in later)

### CODE REVIEW

File: Apps/Contoso.Android.Puppet/Properties/AndroidManifest.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.microsoft.appcenter.xamarin.puppet" android:versionCode="61" android:versionName="2.3.1-SNAPSHOT" android:installLocation="auto">
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="com.microsoft.appcenter.xamarin.puppet" android:versionCode="61" android:versionName="2.4.1-SNAPSHOT" android:installLocation="auto">
```
Comment (guperrot) [inline Apps/Contoso.Android.Puppet/Properties/AndroidManifest.xml]: That would be 2.4.0 and we usually update version in a separate PR. Any chance we can revert those?

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs [hunk 2]
```diff
@@ -21,14 +21,32 @@ public interface IClearCrashClick
         void ClearCrashButton();
     }
 
+    enum PlatformType
+    {
+        Uwp,
```
Comment (guperrot) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs]: Xamarin.Forms has that already, please use https://docs.microsoft.com/en-us/xamarin/xamarin-forms/platform/device

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs [hunk 3]
```diff
@@ -64,7 +82,18 @@ protected override void OnStart()
                 Distribute.SetApiUrl("https://api-gateway-core-integration.dev.avalanch.es/v0.1");
                 Auth.SetConfigUrl("https://config-integration.dev.avalanch.es");
                 Data.SetTokenExchangeUrl("https://token-exchange-mbaas-integration.dev.avalanch.es/v0.1");
-                AppCenter.Start($"uwp={UwpKey};android={AndroidKey};ios={IosKey}", typeof(Analytics), typeof(Crashes), typeof(Distribute), typeof(Auth), typeof(Data));
+
+                // Default to B2C
+                IReadOnlyDictionary<PlatformType, string> AppSecrets = B2CAuthAppSecrets;
```
Comment (guperrot) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs]: This is a local variable, use `var appSecrets`

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs [hunk 4]
```diff
@@ -90,6 +97,18 @@ async void UpdateAuthEnabled(object sender, ToggledEventArgs e)
             await Auth.SetEnabledAsync(e.Value);
         }
 
+        async void ChangeAuthType(object sender, EventArgs e)
+        {
+            int newSelectionCandidate = this.AuthTypePicker.SelectedIndex;
```
Comment (guperrot) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/ModulePages/OthersContentPage.xaml.cs]:
```suggestion
            var newSelectionCandidate = this.AuthTypePicker.SelectedIndex;
```

File: Apps/Contoso.Android.Puppet/Properties/AndroidManifest.xml (hunk 1 above)
Comment (micarls) [reply Apps/Contoso.Android.Puppet/Properties/AndroidManifest.xml]: Sure, will do

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs [hunk 5]
```diff
@@ -64,7 +82,18 @@ protected override void OnStart()
...
+
+                // If user has selected another Auth Type, override the secret dictionary accordingly
```
Comment (achocron) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs]: nit
```suggestion
                // If user has selected another Auth Type, override the secret dictionary accordingly.
```

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs [hunk 6]
```diff
@@ -115,6 +106,20 @@ protected override void OnStart()
...
+        static IReadOnlyDictionary<string, string> GetAppSecretDictionary()
+        {
+            // If user has selected another Auth Type, override the secret dictionary accordingly.
+            var persistedAuthType = AuthTypeUtils.GetPersistedAuthType();
+            switch (persistedAuthType)
+            {
+                case AuthType.AAD:
+                    return AADAuthAppSecrets;
+                default:
+                    // Default to B2C.
```
Comment (guperrot) [inline Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs]: can self document default case by using fallthrough

```
case AuthType.B2C:
default:
```

No need for a comment that way.

Comment (micarls) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet/App.xaml.cs]: Good call, thanks

File: Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.Droid/Properties/AndroidManifest.xml [hunk 7]
```diff
@@ -10,6 +10,14 @@
 				<data android:host="auth" android:scheme="msalbff0949b-7970-439d-9745-92cdc59b10fe" />
 			</intent-filter>
 		</activity>
+		<activity android:name="com.microsoft.identity.client.BrowserTabActivity">
+			<intent-filter>
```
[the remark this answers is not in the document]
Comment (micarls) [reply Apps/Contoso.Forms.Puppet/Contoso.Forms.Puppet.Droid/Properties/AndroidManifest.xml]: Fixed after testing, thanks.

**Comment by micarls** [comment]
Local app test passes now (IOS tested without auth, because IOS SDK auth support not available yet)

**Pull request merged**